### ISSUE TITLE
discussion/possible improvement: a finer control of image fetching during redeploy

### ISSUE THREAD

**Issue by fmobus** [swarmpit/swarmpit#161]
A bit of background first: I have a swarmpit managing my docker swarm, and wanted to integrate it with my CI setup (bitbucket pipelines), so that my staging environment is automatically updated after each successful build.

I spent a couple of days attempting to programatically tell swarmpit to redeploy a service with a new image. The idea was to:

- create (once) a service with image `my-registry/my-image:staging` (at this point, service runs with whatever image is `latest`, e.g. `abc1234`)
- have my CI push a new image (say, `def5678`) to the `staging` tag
- send a simple curl request to `/services/my-service/redeploy` endpoint (authentication worked ok)
- wait until new image is running (by hitting `/services/my-service` and testing `$.repository.imageId`
- build is green if image `def5678` is running within X seconds/minutes/whatever

Unfortunately, I couldn't get it to work properly, as in, the json of `/services/my-service` would never change after a redeploy. The json would properly show the `updating` statuses, but I never saw the `imageId` field change. At first, I thought this was some local swarm issue or something, but as it turns out, upon entering the docker manager host and inspecting all the services and tasks, the new image (`def5678`) was actually there, and it was happily running with no issues.

Additionally, I should add that, at first, I attempted to have my staging environment bound to the `latest` tag. This would produce yet another behaviour: nothing was fetched upon redeploy command, and the same image would remain deployed.

In short, my question is:
- is it a bad idea to try to issue commands to the json api like this?
- is the redeploy command supposed to refetch images from registry every time?
- is the `latest` tag treated differently somehow with regards to redeploys?
- should I be looking for another tool for this kind of usage?

Sorry if this issue looks nothing like a proper bug report.

thanks in advance

**Comment by nohaapav** [comment]
@fmobus Hi, 

Regarding `.repository.imageId` .. These are metadata for swarmpit autoredeploy. Basically if autoredeploy flag is on we check if current imageId is different from imageId in registry -> If yes, we redeploy service automatically with newest image. This value (`.repository.imageId`) is not updated on manual redeploy. Redeploy is basically swarm force command. It should fetch always new version of image if any. 

There shouldn't be any differences between latest/custom tag. Actually i've just tried to redeploy service with latest tag manually and it worked just fine (service was updated to use newest image).

Steps to reproduce:

1.Create service with latest tag from custom registry
2.Change service & push changed image to registry
3.Press service redeploy button.

Could you provide step by step what you're exactly doing to redeploy service manually:
Please use following format so we can debug your case:
https://github.com/swarmpit/swarmpit/blob/master/CONTRIBUTING.md#reporting-issues

Pavol

**Comment by nohaapav** [comment]
Should be improved in https://github.com/swarmpit/swarmpit/releases/tag/1.3.1
We're using now repo digest, same as docker.
